File: scripts/check-featured-images.mjs
#!/usr/bin/env node
/**
 * Playwright test: visits / and /byd across mobile/tablet/desktop and asserts
 * every featured model card renders a real WebP srcset (480w/800w/1280w) with
 * a loaded natural size — i.e. no card falls back to the Zap placeholder icon.
 *
 * Run: node scripts/check-featured-images.mjs
 * Requires: dev server on http://localhost:8080 and playwright installed.
 */
import { chromium } from "playwright";

const BASE = process.env.BASE_URL ?? "http://localhost:8080";
const ROUTES = ["/", "/byd"];
const BREAKPOINTS = [
  { name: "mobile", width: 390, height: 844 },
  { name: "tablet", width: 820, height: 1180 },
  { name: "desktop", width: 1440, height: 900 },
];
const EXPECTED_WIDTHS = ["480w", "800w", "1280w"];

const failures = [];

function fail(msg) {
  failures.push(msg);
  console.error("✗", msg);
}

const browser = await chromium.launch({ headless: true });
try {
  for (const bp of BREAKPOINTS) {
    const ctx = await browser.newContext({ viewport: { width: bp.width, height: bp.height } });
    const page = await ctx.newPage();

    for (const route of ROUTES) {
      const url = `${BASE}${route}`;
      const label = `[${bp.name} ${route}]`;
      await page.goto(url, { waitUntil: "networkidle" });
      // Force lazy images to load.
      await page.evaluate(async () => {
        window.scrollTo(0, document.body.scrollHeight);
        await new Promise((r) => setTimeout(r, 400));
        window.scrollTo(0, 0);
      });
      await page.waitForLoadState("networkidle");

      const cards = await page.$$eval("a[href^='/byd/'], .group", (nodes) =>
        nodes
          .map((n) => n.querySelector("img"))
          .filter((img) => img && /\/assets\/models\//.test(img.currentSrc || img.src))
          .map((img) => ({
            src: img.currentSrc || img.src,
            srcset: img.getAttribute("srcset") ?? "",
            alt: img.alt,
            naturalWidth: img.naturalWidth,
            complete: img.complete,
            hasPlaceholder: !!img.closest(".group")?.querySelector("svg.lucide-zap"),
          })),
      );

      if (cards.length === 0) {
        fail(`${label} no featured model <img> found`);
        continue;
      }

      for (const c of cards) {
        const ctx = `${label} alt="${c.alt}"`;
        if (!/\.webp/i.test(c.src)) fail(`${ctx} src is not webp: ${c.src}`);
        if (!c.complete || c.naturalWidth === 0) fail(`${ctx} image did not load (naturalWidth=${c.naturalWidth})`);
        for (const w of EXPECTED_WIDTHS) {
          if (!c.srcset.includes(w)) fail(`${ctx} srcset missing ${w}: ${c.srcset}`);
        }
      }

      // Assert no placeholder Zap icon rendered inside any card image slot.
      const placeholderCount = await page.locator(".group .aspect-\\[16\\/10\\] > .grid svg.lucide-zap").count();
      if (placeholderCount > 0) fail(`${label} ${placeholderCount} card(s) fell back to Zap placeholder`);

      console.log(`✓ ${label} ${cards.length} card(s) OK`);
    }

    await ctx.close();
  }
} finally {
  await browser.close();
}

if (failures.length) {
  console.error(`\n${failures.length} failure(s)`);
  process.exit(1);
}
console.log("\nAll featured model images verified.");

File: scripts/check-featured-images.py
#!/usr/bin/env python3
"""Playwright test: visits / and /byd across mobile/tablet/desktop and asserts
every featured model card renders a real WebP srcset (480w/800w/1280w) with a
loaded natural size — i.e. no card falls back to the Zap placeholder icon.

Usage: python3 scripts/check-featured-images.py
Requires the dev server on http://localhost:8080 (override with BASE_URL) and
playwright (pre-installed in the Lovable sandbox).
"""
import asyncio
import os
import sys
from playwright.async_api import async_playwright

BASE = os.environ.get("BASE_URL", "http://localhost:8080")
ROUTES = ["/", "/byd", "/compare"]
BREAKPOINTS = [
    ("mobile", 390, 844),
    ("tablet", 820, 1180),
    ("desktop", 1440, 900),
]
EXPECTED_WIDTHS = ["480w", "800w", "1280w"]

CARD_EVAL = """
() => {
  const imgs = Array.from(document.querySelectorAll('img'))
    .filter(i => /\\b480w\\b/.test(i.getAttribute('srcset') || ''));
  return imgs.map(i => ({
    src: i.currentSrc || i.src,
    srcset: i.getAttribute('srcset') || '',
    alt: i.alt,
    naturalWidth: i.naturalWidth,
    complete: i.complete,
  }));
}
"""

async def main():
    failures = []

    def fail(msg):
        failures.append(msg)
        print("FAIL", msg, file=sys.stderr)

    async with async_playwright() as pw:
        browser = await pw.chromium.launch(headless=True)
        try:
            for name, w, h in BREAKPOINTS:
                ctx = await browser.new_context(viewport={"width": w, "height": h})
                page = await ctx.new_page()
                for route in ROUTES:
                    label = f"[{name} {route}]"
                    await page.goto(f"{BASE}{route}", wait_until="networkidle")
                    await page.evaluate(
                        "async () => { window.scrollTo(0, document.body.scrollHeight);"
                        " await new Promise(r => setTimeout(r, 400)); window.scrollTo(0, 0); }"
                    )
                    await page.wait_for_load_state("networkidle")

                    cards = await page.evaluate(CARD_EVAL)
                    if not cards:
                        fail(f"{label} no featured model <img> found")
                        continue

                    for c in cards:
                        ctxlbl = f'{label} alt="{c["alt"]}"'
                        if not c["src"]:
                            fail(f"{ctxlbl} img has no src")
                        if not c["complete"] or c["naturalWidth"] == 0:
                            fail(f"{ctxlbl} image did not load (naturalWidth={c['naturalWidth']})")
                        for want in EXPECTED_WIDTHS:
                            if want not in c["srcset"]:
                                fail(f"{ctxlbl} srcset missing {want}: {c['srcset']}")

                    # No need to check for Zap placeholder separately: the img/placeholder
                    # branches are exclusive, so a valid srcset on every card guarantees
                    # no fallback was rendered.

                    print(f"OK {label} {len(cards)} card(s)")
                await ctx.close()
        finally:
            await browser.close()

    if failures:
        print(f"\n{len(failures)} failure(s)", file=sys.stderr)
        sys.exit(1)
    print("\nAll featured model images verified.")

asyncio.run(main())
